frontend/pages/labels/EditLabelPage/EditLabelPage.tsx

```diff
@@ -0,0 +1,139 @@
+import React, { useContext } from "react";
+import { useQuery } from "react-query";
+import { RouteComponentProps } from "react-router";
+import { AxiosError } from "axios";
+
+import PATHS from "router/paths";
+import labelsAPI, { IGetLabelResonse } from "services/entities/labels";
+import hostAPI from "services/entities/hosts";
+import { DEFAULT_USE_QUERY_OPTIONS } from "utilities/constants";
+import { ILabel } from "interfaces/label";
+import { IHost } from "interfaces/host";
+import { NotificationContext } from "context/notification";
+
+import MainContent from "components/MainContent";
+import Spinner from "components/Spinner";
+import DataError from "components/DataError";
+
+import DynamicLabelForm from "../components/DynamicLabelForm";
+import ManualLabelForm from "../components/ManualLabelForm";
+import { IDynamicLabelFormData } from "../components/DynamicLabelForm/DynamicLabelForm";
+import { IManualLabelFormData } from "../components/ManualLabelForm/ManualLabelForm";
+
+const baseClass = "edit-label-page";
+
+interface IEditLabelPageRouteParams {
+  label_id: string;
+}
+
+type IEditLabelPageProps = RouteComponentProps<
+  never,
+  IEditLabelPageRouteParams
+>;
+
+const EditLabelPage = ({ routeParams, router }: IEditLabelPageProps) => {
+  const { renderFlash } = useContext(NotificationContext);
+
+  const labelId = parseInt(routeParams.label_id, 10);
+
+  const {
+    data: label,
+    isLoading: isLoadingLabel,
+    isError: isErrorLabel,
+  } = useQuery<IGetLabelResonse, AxiosError, ILabel>(
+    ["label", labelId],
+    () => labelsAPI.getLabel(labelId),
+    {
+      ...DEFAULT_USE_QUERY_OPTIONS,
+      select: (data) => data.label,
+    }
+  );
+
+  // TODO: clean this up when API allows getting hosts by
+  // host ids in a single request. We need to make another request when
+  // the lable is manual to get the host data for the targeted hosts.
+  const {
+    data: targetedHosts,
+    isLoading: isLoadingHosts,
+    isError: isErrorHosts,
+  } = useQuery<{ host: IHost }[], AxiosError, IHost[]>(
+    ["hosts"],
+    () => {
+      return hostAPI.getHosts(label?.host_ids ?? []);
+    },
+    {
+      ...DEFAULT_USE_QUERY_OPTIONS,
+      select: (res) => res.map((host) => host.host),
+      enabled: label?.label_membership_type === "manual",
+    }
+  );
+
+  const onCancelEdit = () => {
+    router.goBack();
+  };
+
+  const onUpdateLabel = async (
+    formData: IDynamicLabelFormData | IManualLabelFormData
+  ) => {
+    try {
+      const res = await labelsAPI.update(labelId, formData);
+      router.push(PATHS.MANAGE_HOSTS_LABEL(res.label.id));
+      renderFlash("success", "Label updated successfully.");
+    } catch {
+      renderFlash("error", "Couldn't edit label. Please try again.");
+    }
+  };
+
+  const renderContent = () => {
+    if (isLoadingLabel || isLoadingHosts) {
+      return <Spinner />;
+    }
+
+    if (isErrorLabel || isErrorHosts) {
+      return <DataError />;
```

Comment:
@rachaelshaw @marko-lisica how do we want to handle errors, like if we go to the URL /labels/99999999999, it is handled differently than queries/999999999999 or policies/9999999999

Reply:
@RachelElysia If I see correctly, if we go to /labels/{non-existing-id} it opens label editor, and for queries and policies we return 404 page. I think we should probably be consistent and return 404 page.

cc @rachaelshaw